modules/ethereum-contract/builtin/src/lib.rs

```diff
@@ -45,35 +49,266 @@ pub struct Header {
+						.parse()
+						.unwrap(),
+					digest,
+				}
+				.encode()
+			)
+			.unwrap(),
+			Header {
+				hash: "3dfebb280bd87a4640f89d7f2adecd62b88148747bff5b63af6e1634ee37a56e"
+					.parse()
+					.unwrap(),
+				parent_hash: "c0ac300d4005141ea690f3df593e049739c227316eb7f05052f3ee077388b68b"
+					.parse()
+					.unwrap(),
+				number: 8,
+				signal: Some(ValidatorsSetSignal {
+					delay: 8,
+					validators: authorities.encode(),
+				}),
+			},
+		);
+	}
+
+	/// Number of the example block with justification.
+	const EXAMPLE_JUSTIFIED_BLOCK_NUMBER: u32 = 8;
+	/// Hash of the example block with justification.
+	const EXAMPLE_JUSTIFIED_BLOCK_HASH: &'static str =
+		"a2f45892db86b2ad133ce57d81b7e4375bb7035ce9883e6b68c358164f343775";
+	/// Id of authorities set that have generated example justification. Could be computed by tracking
+	/// every set change in canonized headers.
+	const EXAMPLE_AUTHORITIES_SET_ID: u64 = 0;
+	/// Encoded authorities set that has generated example justification. Could be fetched from `ScheduledChange`
+	/// digest of the block that has scheduled this set OR by calling `GrandpaApi::grandpa_authorities()` at
+	/// appropriate block.
+	const EXAMPLE_AUTHORITIES_SET: &'static str = "1488dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0ee0100000000000000d17c2d7823ebf260fd138f2d7e27d114c0145d968b5ff5006125f2414fadae690100000000000000439660b36c6c03afafca027b910b4fecf99801834c62a5e6006f27d978de234f01000000000000005e639b43e0052c47447dac87d6fd2b6ec50bdd4d0f614e4299c665249bbd09d901000000000000001dfe3e22cc0d45c70779c1095f7489a8ef3cf52d62fbd8c2fa38c9f1723502b50100000000000000";
+	/// Example justification. Could be fetched by calling 'chain_getBlock' RPC.
+	const EXAMPLE_JUSTIFICATION: &'static str = "2600000000000000a2f45892db86b2ad133ce57d81b7e4375bb7035ce9883e6b68c358164f3437750800000010a2f45892db86b2ad133ce57d81b7e4375bb7035ce9883e6b68c358164f34377508000000d66b4ceb57ef8bcbc955071b597c8c5d2adcfdbb009c73f8438d342670fdeca9ac60686cbd58105b10f51d0a64a8e73b2e5829b2eab3248a008c472852130b00439660b36c6c03afafca027b910b4fecf99801834c62a5e6006f27d978de234fa2f45892db86b2ad133ce57d81b7e4375bb7035ce9883e6b68c358164f34377508000000f5730c14d3cd22b7661e2f5fcb3139dd5fef37f946314a441d01b40ce1200ef70d810525f23fd278b588cd67473c200bda83c338c407b479386aa83798e5970b5e639b43e0052c47447dac87d6fd2b6ec50bdd4d0f614e4299c665249bbd09d9a2f45892db86b2ad133ce57d81b7e4375bb7035ce9883e6b68c358164f34377508000000c78d6ec463f476461a695b4791d30e7626d16fdf72d7c252c2cad387495a97e8c2827ed4d5af853d6e05d31cb6fb7438c9481a7e9c6990d60a9bfaf6a6e1930988dc3417d5058ec4b4503e0c12ea1a0a89be200fe98922423d4334014fa6b0eea2f45892db86b2ad133ce57d81b7e4375bb7035ce9883e6b68c358164f3437750800000052b4fc52d430286b3e2d650aa6e01b6ff4fae8b968893a62be789209eb97ee6e23780d3f5af7042d85bb48f1b202890b22724dfebce138826f66a5e00324320fd17c2d7823ebf260fd138f2d7e27d114c0145d968b5ff5006125f2414fadae6900";
```

Comment:
Thanks for the explanations!

Is there a way to actually generate that as well? How involved would it be?

Reply:
There's no way - it's encoded `GrandpaJustification`, struct with private fields.

Reply:
Could we just copy paste the code on our side? Sorry for pushing this so hard, but I remember the experience of debugging smart contracts and JSON tests in Ethereum, where such encoded hexes where pretty common. Within few months I bet we will completely forget the knowledge where this hexes come from and what were the assumptions and having a way to alter the parameters easily is going to be crucial.
Also if we want to add more tests in the future, we need a way to easily generate different edge cases that we need - with hexes the path of least resistance is to not add a test at all, cause the cost of generating the blob is super high if you don't know how to do it.

Reply:
OK, I think we may split your comment into two parts.

>  I bet we will completely forget the knowledge where this hexes come from and what were the assumptions and having a way to alter the parameters easily is going to be crucial

I agree that simple hex literals weren't good enough - now we have a comments on how to update this. So if anything breaks, the only thing you need is to get block (hash and number) and justification from the node - that is well commented (I hope). Justification is currently opaque to the builtin and I'm not testing if **justification generation ** and **justification verification** works correctly here. I don't care whether block hash is simply encoded `int` and justification is simply two encoded `int`s which sum is the block hash, or it is some multisig structure. The only thing I'm testing here is that if the contract stores valid (from `sc_finality_grandpa` point of view) values of **header**, **validators set** and **justification**, then the block is finalized. And that it isn't finalized if contract stores some garbage.

> cause the cost of generating the blob is super high if you don't know how to do it

Please see comments above constants to see how to get these blobs. I don't think that cost of calling two RPC methods on first block with justification is that high. I'd be happy to fix comments if you have anything to add.

Reply:
> Justification is currently opaque to the builtin and I'm not testing if **justification generation ** and justification verification works correctly her

I fully understand that. We should definitely make sure that we are able to parse not the things that are generated by our code, but rather whatever substrate chain generates.

> The only thing I'm testing here is that if the contract stores valid (from sc_finality_grandpa point of view) values of header, validators set and justification, then the block is finalized. And that it isn't finalized if contract stores some garbage.

Totally agree with this too. However my main point with high cost of generating this is that it's not easy to control it. Like imagine you want to add more tests with some specific validator set (say one validator or even number of validators), or some peculiarity with Justification that has to be handled differently. In such case you would need to run a specific substrate chain, wait until it generates a particular justification you need and only then retrieve it over the RPC.

That said, I'm really far from blocking the PR and pushing you to implement that. I'm just seeing this from a bystander perspective, didn't really have a chance to work on the code directly to see if this is a really issue or just my imagination :)